### PR TITLE
Fewshot learning predict proba

### PR DESCRIPTION
**Pull request by yongxinw**
*Issue #, if available:*

*Description of changes:*
Adding predict proba for fewshot svm learner

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/utils/few_shot_learning.py [hunk 1]
```diff
@@ -91,11 +91,33 @@ def predict(self, data, as_pandas: Optional[bool] = False):
         features = self.extract_embedding(data)
 
         preds = self.clf.predict(features)
+        # import ipdb
+
+        # ipdb.set_trace()
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/few_shot_learning.py:96]: Need to remove the debugging code.

File: multimodal/src/autogluon/multimodal/utils/few_shot_learning.py [hunk 2]
```diff
@@ -91,11 +91,33 @@ def predict(self, data, as_pandas: Optional[bool] = False):
...
 
+    def predict_proba(self, data, as_pandas: Optional[bool] = False, as_multiclass: Optional[bool] = True):
+        if not self._fit_called and not self._model_loaded:
+            warnings.warn(
+                "Neither .fit() nor .load() is not invoked. Unexpected predictions may occur. Please consider calling .fit() or .load() before .predict()"
+            )
+
+        features = self.extract_embedding(data)
+        preds = self.clf.decision_function(features)
+        probs = np.exp(preds) / np.sum(np.exp(preds + 1e-8), axis=1, keepdims=True)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/few_shot_learning.py:109]: Refer to https://github.com/autogluon/autogluon/blob/master/multimodal/src/autogluon/multimodal/predictor.py#L2288

File: multimodal/src/autogluon/multimodal/utils/few_shot_learning.py [hunk 3]
```diff
@@ -91,11 +91,33 @@ def predict(self, data, as_pandas: Optional[bool] = False):
...
+
+        if not as_multiclass:
+            probs = probs.max(axis=1)
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/few_shot_learning.py:112]: Refer to https://github.com/autogluon/autogluon/blob/master/multimodal/src/autogluon/multimodal/predictor.py#L2290-L2292

File: multimodal/src/autogluon/multimodal/utils/few_shot_learning.py [hunk 4]
```diff
@@ -91,11 +91,33 @@ def predict(self, data, as_pandas: Optional[bool] = False):
...
+
+        if as_pandas:
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/few_shot_learning.py:114]: Refer to https://github.com/autogluon/autogluon/blob/master/multimodal/src/autogluon/multimodal/predictor.py#L2294

File: multimodal/tests/unittests/others_2/test_fewshot_predictor.py
Comment (zhiqiangdon) [inline multimodal/tests/unittests/others_2/test_fewshot_predictor.py:1]: Can we use the similar test logic as `test_predictor` ?https://github.com/autogluon/autogluon/blob/master/multimodal/tests/unittests/predictor/test_predictor.py#L252-L282
`test_predictor` uses two utils (`verify_predictor_save_load` and `verify_realtime_inference`) and tests continuous training.

**Review by zhiqiangdon — COMMENTED** [review]
Need to resolve the conflict with the master branch.

File: multimodal/src/autogluon/multimodal/utils/few_shot_learning.py (hunk 1 above)
Comment (yongxinw) [reply multimodal/src/autogluon/multimodal/utils/few_shot_learning.py:96]: Done

File: multimodal/tests/unittests/others_2/test_fewshot_predictor.py [hunk 5]
```diff
@@ -1,7 +1,43 @@
+import numpy as np
+
 from autogluon.multimodal.utils.few_shot_learning import FewShotSVMPredictor
 from autogluon.multimodal.utils.misc import shopee_dataset
 
 
+def verify_evaluate(test_data, predictor):
+    results = predictor.evaluate(test_data)
+
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/others_2/test_fewshot_predictor.py:9]: Can we reuse `verify_predictor_save_load` (https://github.com/autogluon/autogluon/blob/master/multimodal/tests/unittests/predictor/test_predictor.py#L46)? We may need to merge the tests after the learner refactor.

File: multimodal/tests/unittests/others_2/test_fewshot_predictor.py [hunk 6]
```diff
@@ -1,7 +1,43 @@
...
+    acc = (preds == test_data["label"].to_numpy()).sum() / len(test_data)
+
+    preds2 = predictor.predict(test_data)
+    assert len(preds2) == len(test_data)
+    assert (preds == preds2).all()
+    return preds
+
+
 def test_fewshot_fit_predict():
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/others_2/test_fewshot_predictor.py:41]: Can we also test realtime inference (https://github.com/autogluon/autogluon/blob/master/multimodal/tests/unittests/predictor/test_predictor.py#LL263C5-L263C30)?

File: multimodal/tests/unittests/others_2/test_fewshot_predictor.py [hunk 7]
```diff
@@ -23,17 +59,13 @@ def test_fewshot_fit_predict():
...
-    acc = (preds == test_data["label"].to_numpy()).sum() / len(test_data)
+    verify_evaluate(test_data, predictor)
 
-    preds2 = predictor.predict(test_data)
-    assert len(preds2) == len(test_data)
-    assert (preds == preds2).all()
+    verify_predict_predict_proba(test_data, predictor)
 
-    results = predictor.evaluate(test_data)
+    verify_predict_as_pandas_multiclass(test_data, predictor)
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/others_2/test_fewshot_predictor.py:68]: Continuous training seems also missing (https://github.com/autogluon/autogluon/blob/master/multimodal/tests/unittests/predictor/test_predictor.py#L275-L283).

**Review by zhiqiangdon — COMMENTED** [review]
We may need to align with `test_predictor` (https://github.com/autogluon/autogluon/blob/master/multimodal/tests/unittests/predictor/test_predictor.py#L252-L283) to facilitate future learner refactor. After the refactor, few shot predictor will reuse the MultiModalPredictor APIs, so the tests should also be compatible.

**Review by zhiqiangdon — APPROVED** [review]
LGTM. Thanks for aligning the tests!

**Pull request merged**